src/main/scala/com/fulcrumgenomics/commons/async/AsyncWriterPool.scala

```diff
@@ -0,0 +1,188 @@
+  * close all open writers, shutdown the internal thread pool, and rethrow the first exception (if
+  * any) encountered when closing writers.
+  *
+  * @param threads the number of threads to use to write items.
+  */
+class AsyncWriterPool(threads: Int) extends Closeable {
+  require(threads >= 1, "Cannot create a pool with fewer than 1 thread(s).")
+  private val executor   = Executors.newFixedThreadPool(threads)
+  private val writers    = new CopyOnWriteArrayList[PooledWriter[_]]()
+  private val poolClosed = new AtomicBoolean(false)
+  Range(0, threads).foreach(_ => executor.submit(new Drainer))
+
+  /**
+    * Wraps a [[Writer[A]]] in the context of an [[AsyncWriterPool]] to provide asynchronous operations.
+    * Calls to [[write()]] and related methods place items into the queue, and if necessary, schedule
+    * the queue for draining using the [[executor]] in the enclosing [[AsyncWriterPool]].
+    */
+  class PooledWriter[A] private[AsyncWriterPool] (private val writer: Writer[A],
+                                                  private val queue: BlockingQueue[A]) extends Writer[A] {
+    /** A reference to an exception thrown during an asynchronous operation. */
+    private[AsyncWriterPool] val throwable: AtomicReference[Throwable] = new AtomicReference(null)
+
+    /** True if this writer is enqueued to be drained, false otherwise. */
+    private[AsyncWriterPool] val draining: AtomicBoolean = new AtomicBoolean(false)
+
+    /** Drains the queue to the underlying writer. Returns the count of items drained. */
+    private[AsyncWriterPool] def drain(): Int = {
+      this.synchronized {
+        var item: A = queue.poll()
+        var count = 0
+        while (item != null) {
+          writer.write(item)
+          count += 1
+          item = queue.poll()
+        }
+
+        count
+      }
+    }
+
+    /** True if a close() method hs been called, even if closing() hasn't happened yet. */
+    private[AsyncWriterPool] val writerCloseRequested: AtomicBoolean = new AtomicBoolean(false)
+
+    /** True if a close() method has been called, false otherwise. */
+    private[AsyncWriterPool] val writerClosed: AtomicBoolean = new AtomicBoolean(false)
+
+    /** Checks to see if an exception has been raised asynchronously, and rethrows it on the current thread. */
+    private def checkAndRaise(): Unit = {
+      val t = throwable.getAndSet(null)
+      if (t != null) throw t
+    }
+
+    /** Writes an individual item. May block if a bounded queue is in use. */
+    override def write(item: A): Unit = {
+      checkAndRaise()
+      if (writerCloseRequested.get()) throw new IllegalStateException("Cannot write to closed() writer.")
+      queue.put(item)
+    }
+
+    /** Closes the writer _synchronously_ on the current thread. */
+    override def close(): Unit = {
+      checkAndRaise()
+      writerCloseRequested.set(true)
+      AsyncWriterPool.this.writers.remove(this)
+      if (!writerClosed.getAndSet(true)) {
+        drain()
```

Comment:
Should we check that this writer isn't draining before calling `drain()`?  The `Drainer`s you created could be draining.  I suppose that because `drain()` is synchronized it shouldn't matter.